### PR TITLE
Feature vector and saliency map dumping

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
Changes to dump feature vector and source features for saliecy map for Object detection, Counting and Rotated Detection

### CODE REVIEW

**Comment by GalyaZalesskaya** [comment]
run ote_sdk tests

File: mmdet/models/detectors/single_stage.py [hunk 1]
```diff
@@ -110,15 +110,21 @@ def simple_test(self, img, img_metas, rescale=False, postprocess=True):
         with no_nncf_trace():
             bbox_results = \
                 self.bbox_head.get_bboxes(*outs, img_metas, self.test_cfg, False)
+
+        feature_map = outs[0]  # classification head output
+        pooled_features = [torch.nn.functional.adaptive_avg_pool2d(feat_map, (1, 1))
+                           for feat_map in x]  # x - backbone+neck output
+        feature_vector = torch.cat(pooled_features, dim=1)
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/single_stage.py]: Is it computed all the time? Should it be?

**Comment by GalyaZalesskaya** [comment]
run ote-test

**Comment by Ilya-Krylov** [comment]
please have a look at tests

**Comment by GalyaZalesskaya** [comment]
@Ilya-Krylov  OTE tests failed, because these changes require one-time changes in OTE rep, so it doesn't work with the OTE develop branch. For this reason I've run a tests on the [OTE side](https://ci-ote.iotg.sclab.intel.com/job/ote/job/ote-test/1630/), putting this branch as mmdetection backend. But it also failed, so I'm looking into this.

File: mmdet/apis/export.py [hunk 2]
```diff
@@ -55,6 +55,7 @@ def patch_nncf_model_for_alt_ssd_export(model):
 def export_to_onnx(model,
                    data,
                    export_name,
+                   output_names,
```
Comment (druzhkov-paul) [inline mmdet/apis/export.py]: Do you pass in here anything apart from `['boxes', 'labels']`? If not, why do you need this extra argument?

File: mmdet/models/detectors/single_stage.py (hunk 1 above)
Comment (GalyaZalesskaya) [reply mmdet/models/detectors/single_stage.py]: Yes, it's computed all the time. If we don't combine X in the feature vector, we would need to pass all this backbone output as a list of tensors, that will add a lot of unnamed blobs in ONNX output_names. And these unnamed blobs will be hard to handle and create feature vector from that

File: mmdet/apis/export.py (hunk 2 above)
Comment (GalyaZalesskaya) [reply mmdet/apis/export.py]: Actually, you're right, we don't need it and probably it provokes the problem with CI. I'll remove it

File: mmdet/models/detectors/single_stage.py (hunk 1 above)
Comment (druzhkov-paul) [reply mmdet/models/detectors/single_stage.py]: Pooling and concatenation is fine, I was concerned about the fact that these feature maps and vectors are computed all the time. Is it fine in terms of extra memory required during the dataset inference?

Comment (GalyaZalesskaya) [reply mmdet/models/detectors/single_stage.py]: I measured the GPU memory consumption with this calculation/dumping and without it. I measured it on the demo with 1 image and on the whole dataset evaluation. The memory consumption is the same in all cases (1645mb). And I didn't catch any OOO errors during e2e procedure

Comment (druzhkov-paul) [reply mmdet/models/detectors/single_stage.py]: You mean the GPU memory or RAM? What is OOO error? 😄

Comment (GalyaZalesskaya) [reply mmdet/models/detectors/single_stage.py]: I meant GPU memory, should I measure RAM? Yes, I meant OOM (Out of Memory ) error

**Review by druzhkov-paul — APPROVED** [review]

File: mmdet/models/detectors/two_stage.py [hunk 3]
```diff
@@ -204,9 +204,15 @@ def simple_test(self,
...
 
-        return self.roi_head.simple_test(
+        out = self.roi_head.simple_test(
             x, proposal_list, img_metas, rescale=rescale, postprocess=postprocess)
 
+        pooled_features = [torch.nn.functional.adaptive_avg_pool2d(feat_map, (1, 1))
+                           for feat_map in x]  # x - backbone+neck output
+        feature_vector = torch.cat(pooled_features, dim=1)
+
+        return out, feature_vector
```
Comment (goodsong81) [inline mmdet/models/detectors/two_stage.py:214]: @eugene123tw The semantic of 2nd output of this function is different from above (mmdet/models/detectors/single_stage.py). I think this need to be fixed.

**Review by goodsong81 — CHANGES_REQUESTED** [review]

File: mmdet/models/detectors/single_stage.py [hunk 4]
```diff
@@ -110,15 +110,21 @@ def simple_test(self, img, img_metas, rescale=False, postprocess=True):
         with no_nncf_trace():
             bbox_results = \
                 self.bbox_head.get_bboxes(*outs, img_metas, self.test_cfg, False)
+            pooled_features = [torch.nn.functional.adaptive_avg_pool2d(feat_map, (1, 1))
+                               for feat_map in x]  # x - backbone+neck output
+            feature_vector = torch.cat(pooled_features, dim=1)
+
+        feature_map = outs[0]  # classification head output
```
Comment (eugene123tw) [inline mmdet/models/detectors/single_stage.py:117]: @goodsong81 I'll also change this part. `outs` is the anchor box prediction not the activation map
